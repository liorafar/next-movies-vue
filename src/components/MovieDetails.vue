<template>
<div class="overlay">
    <div class="movie-details-window">
        <div class="movie-details-container">
            <div class="image-block">
                <img :src="movie.largeimage" alt="test image">
            </div>
            <div class="movie-details">
                <div class="close-movie-details">
                    <h3 class="go-back-button" @click="goBackClicked">X</h3>
                </div>
                <h2 v-html="$t(movie.title)" class="movie-name"></h2>
                <!-- To do: add rating stars -->
                <h3>{{ releasedRatingRuntime }}</h3>
                <p v-html="$t(movie.synopsis)" class="synopsis"></p>
                <a target="_blank" :href="`https://www.imdb.com/title/${movie.imdbid}/`">IMDB</a>
            </div>
        </div>
    </div>
    </div>
</template>

<script>
import movieService from '../services/movie'

export default {
  name: 'MovieDetails',
  async created(){
       try {
        this.movie = await movieService.getMovieById(this.id);
      } catch (error) {
          console.error(`Error: Id: ${error.id}, Exception: ${error.error}`);
      }
  },
  props: ['id'],
  data(){
      return {
        movie: {}
      }
  },
  methods: {
      goBackClicked(){
          this.$root.bus.$emit('goBackClicked');
      }
  },
  computed: {
      releasedRatingRuntime(){
          const runtime = this.movie.runtime ? ` | ${this.movie.runtime}` : '';
          const rating = this.movie.rating ? ` | ${this.movie.rating}` : '';
          return `${this.movie.released}${runtime}${rating}`;
      }
  }
}
</script>

<style lang="scss" scoped>
    .overlay{
        position: fixed;
        height: 100%;
        width: 100%;
        top: 0;
        bottom: 0;
        left: 0;
        right: 0;
        background: rgba(0, 0, 0, 0.5);
        transition: opacity 200ms;
        visibility: visible;
        opacity: 1;
        // this is since the range bar
        // has very high z index which
        // makes it top of the overlay
        z-index: 1000000000000000; 
    }

    .movie-details-window{
        display:absolute;
        flex-direction: column;
        background-color: #050304;
        box-sizing: content-box;
        position: fixed;
        top: 50%;
        left: 50%;
        margin-right: -50%;
        transform: translate(-50%, -50%);
        height: 398px;
        width: 800px;
        overflow-y: auto;
        margin: 0 auto;
        border-radius: 5px;

        

        .movie-details-container {
            display:flex;

            .image-block {
                display: flex;

                img {
                    height: 398px;
                }
            }

            .movie-details{
                display:flex;
                flex-direction: column;
                padding: 50px;

                .close-movie-details{
                    display: flex;
                    height: 20px;
                    width:100%;

                    .go-back-button{
                        position:absolute;
                        right: 15px;
                        top: 10px;
                        color:white;

                        &:hover{
                            cursor: pointer;
                        }
                    }
                }

                h2 {
                    color: #ffffff;
                    font-size: 35px;
                    line-height: 42px;
                }

                h3 {
                    color: #ffffff;
                    font-size: 18px;
                    line-height: 18px;
                }
                
                h2, h3 {
                    margin-top: 0;
                    margin-bottom: 0.5em;
                    font-weight: 500;
                }
                
                .synopsis{
                    color: gray;
                    font-size: 16px;
                    font-weight: 300;
                    line-height: 20px;
                    margin-top: 10px;
                }
            }
        }
    }
    

    
</style>