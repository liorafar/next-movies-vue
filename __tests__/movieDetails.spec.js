import MovieDetails from '../src/components/MovieDetails'
import fetchMock from 'jest-fetch-mock'
import VueTestUtils, { mount, shallowMount, createLocalVue } from '@vue/test-utils'

describe('MovieDetails Component', () => {
    // Inspect the raw component options
    it('has data', () => {
        expect(typeof MovieDetails.data).toBe('function')
    })
})

describe('Mounted MovieDetails', () => {
    let wrapper;

    beforeEach(async () => {
        fetchMock.enableMocks(); // injecting fetch mock due to created hook calling fetch api
        fetchMock.mockResponse(JSON.stringify(['stub for created hook']) , { status: 200 });
  

        wrapper = await shallowMount(MovieDetails, {
            mocks: {
                $t: () => 'batman' // mocking the $t function of VueI18n
            },
            propsData: { 
                id:   1234 // mocking prop
            },         
        })
    })

    it('is instantiated', () => {
        console.log(wrapper.html())
        expect(wrapper).toBeTruthy()
    });

    // rendering test based on parameters
    it("renders released 1995, rating 8.3 and runtime 1h34m as '1995 | 1h34m | 8.3'", async () => {
        wrapper.setData({ movie: {title: "batman", imdbid: "tt0062622", released: 1995, rating: 8.3,  runtime: "1h34m"} })
        await wrapper.vm.$forceUpdate()
        console.log(wrapper.html())
        expect(wrapper.vm.releasedRatingRuntime).toBe("1995 | 1h34m | 8.3");
    })

    it("renders released 1995, no rating and runtime 1h34m as '1995 | 8.3'", async () => {
        wrapper.setData({ movie: {title: "batman", imdbid: "tt0062622", released: 1995, rating: 8.3 } })
        await wrapper.vm.$forceUpdate()
        console.log(wrapper.html())
        expect(wrapper.vm.releasedRatingRuntime).toBe("1995 | 8.3");
    })

    // Snapshot testing
    it("renders released 1995, no rating and no runtime as '1995'", async () => {
        wrapper.setData({ movie: {title: "batman", imdbid: "tt0062622", released: 1995 } })
        await wrapper.vm.$forceUpdate()
        console.log(wrapper.html())
        expect(wrapper.element).toMatchSnapshot()
    })
})