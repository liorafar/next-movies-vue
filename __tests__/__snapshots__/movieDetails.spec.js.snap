// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted MovieDetails renders released 1995, no rating and no runtime as '1995' 1`] = `
<div
  class="overlay"
>
  <div
    class="movie-details-window"
  >
    <div
      class="movie-details-container"
    >
      <div
        class="image-block"
      >
        <img
          alt="test image"
        />
      </div>
       
      <div
        class="movie-details"
      >
        <div
          class="close-movie-details"
        >
          <h3
            class="go-back-button"
          >
            X
          </h3>
        </div>
         
        <h2
          class="movie-name"
        >
          batman
        </h2>
         
        <h3>
          1995
        </h3>
         
        <p
          class="synopsis"
        >
          batman
        </p>
         
        <a
          href="https://www.imdb.com/title/tt0062622/"
          target="_blank"
        >
          IMDB
        </a>
      </div>
    </div>
  </div>
</div>
`;
